overlord/assertstate/assertstate.go

```diff
@@ -1129,3 +1129,78 @@ func TemporaryDB(st *state.State) *asserts.Database {
 	db := cachedDB(st)
 	return db.WithStackedBackstore(asserts.NewMemoryBackstore())
 }
+
+// ValidationSetsFromModel takes in a model and creates a
+// snapasserts.ValidationSets from any validation sets that the model includes.
+func ValidationSetsFromModel(st *state.State, model *asserts.Model, store snapstate.StoreService, offline bool) (*snapasserts.ValidationSets, error) {
+	var sets []*asserts.ValidationSet
+	save := func(a asserts.Assertion) error {
+		if vs, ok := a.(*asserts.ValidationSet); ok {
+			sets = append(sets, vs)
+		}
+
+		if err := Add(st, a); err != nil {
+			if err, ok := err.(*asserts.RevisionError); ok {
+				logger.Noticef("assertion not added due to same or newer revision already present: %d", err.Current)
+				return nil
+			}
+			return err
+		}
+
+		return nil
+	}
+
+	db := DB(st)
+
+	retrieve := func(ref *asserts.Ref) (asserts.Assertion, error) {
+		if offline {
+			return ref.Resolve(db.Find)
+		}
+
+		st.Unlock()
+		defer st.Lock()
+
+		return store.Assertion(ref.Type, ref.PrimaryKey, nil)
+	}
+
+	retrieveSeq := func(ref *asserts.AtSequence) (asserts.Assertion, error) {
+		if offline {
+			return resolveValidationSetAssertion(ref, db)
+		}
+
+		st.Unlock()
+		defer st.Lock()
+
+		return store.SeqFormingAssertion(ref.Type, ref.SequenceKey, ref.Sequence, nil)
+	}
+
+	fetcher := asserts.NewSequenceFormingFetcher(db, retrieve, retrieveSeq, save)
```

Comment:
You could split the creation of the fetcher into a helper, to make the function easier to read, i.e `setupFetcherForValidationSets` or something like that.

Reply:
Talked about why this can't quite work due to closing over the `sets` slice.